### ISSUE TITLE
how to add parameters in user custom function?

### ISSUE THREAD

**Issue by wangbingnan136** [dmlc/dgl#4151]
def edge_udf(edges):
    """
    Parameters
    ----------
    edges : EdgeBatch
        A batch of edges.

    Returns
    -------
    dict[str, tensor]
        The messages or edge features generated. It maps a message/feature name to the
        corresponding messages/features of all edges in the batch. The order of the
        messages/features is the same as the order of the edges in the input argument.
    """

How to write custom function like this

def edge_udf(edges,params):
    """
    Parameters
    ----------
    edges : EdgeBatch
        A batch of edges.

    Returns
    -------
    dict[str, tensor]
        The messages or edge features generated. It maps a message/feature name to the
        corresponding messages/features of all edges in the batch. The order of the
        messages/features is the same as the order of the edges in the input argument.
    """
????
      with g.local_scope():
          g.srcdata['h'] = h # not ndata
          g.edata['w'] = w
          g.update_all(message_func=u_cat_e(params), reduce_func=fn.mean('m', 'h_N'))

          h_N = g.dstdata['h_N']
          h = g.dstdata['feature']
          h_total = torch.cat([h, h_N], dim=1)
          return self.linear(h_total)

**Comment by mufeili** [comment]
If you just want to do `self.linear`, then you can make `edge_udf` a class method like `edge_udf(self, edges)`.

**Comment by luoyoucode** [comment]
I have a heterograph
```
 g = dgl.heterograph({
        ('user', 'follows', 'user'): ([0, 1, 0], [1, 1, 1]),
        ('user', 'attracts', 'user'): ([0, 1, 0], [1, 1, 1]),
    })
    g.nodes['user'].data['h'] = torch.tensor([[1., 2.], [3., 4.]])
```
and  UDF
```
def view_level_sum(nodes):
    v = nodes.data['h'].reshape(1, 1, -1)
    nei = torch.cat((nodes.data['h'].reshape(1, 1, -1), nodes.mailbox['m']), dim=1)
    model = attn_module(nei.shape[2], 5, 3)
    a_ij = model(v, nei).reshape(1,-1,1)
    nodes.mailbox['m'] = nei * a_ij
    return {'h': nodes.mailbox['m'].sum(1)}
```
`attn_module` is NN，based class `nn.module`，which will return a tensor of attention weight `a_ij` like:
```
tensor([[[0.2548],
         [0.2452],
         [0.2548],
         [0.2452]]], grad_fn=<ViewBackward>)
```
In heterograph, I want to set a corresponding`attn_module`  for each corresponding `canonical_etypes`'s UDF, instead of initializing a `attn_module` every time UDF is called.Therefore, I want to add parameters to UDF, as follows
```
def view_level_sum(nodes,model):
    v = nodes.data['h'].reshape(1, 1, -1)
    nei = torch.cat((nodes.data['h'].reshape(1, 1, -1), nodes.mailbox['m']), dim=1)
    a_ij = model(v, nei).reshape(1,-1,1)
    nodes.mailbox['m'] = nei * a_ij
    return {'h': nodes.mailbox['m'].sum(1)}

m1 = attn_module(nei.shape[2], 5, 3)
m2 = attn_module(nei.shape[2], 6, 4)

g.multi_update_all(
        {
            ('user', 'follows', 'user'): (fn.copy_src('h', 'm'), view_level_sum(m1)),
             ('user', 'attracts', 'user'): (fn.copy_src('h', 'm'), view_level_sum(m2)),
        },
        'max')
```
What shall I do?
